packages/transport-webtransport/src/stream.ts

```diff
@@ -1,184 +1,110 @@
-      }
-    }
-    writerClosed = true
-    if (writerClosed && readerClosed) {
-      cleanupStreamFromActiveStreams()
-    }
-  })().catch(() => {
-    log.error('WebTransport failed to cleanup closed stream')
-  });
-
-  (async function () {
-    const err: Error | undefined = await reader.closed.catch((err: Error) => err)
-    if (err != null) {
-      log.error(`WebTransport reader closed unexpectedly: streamId=${streamId} err=${err.message}`)
-    }
-    readerClosed = true
-    if (writerClosed && readerClosed) {
-      cleanupStreamFromActiveStreams()
-    }
-  })().catch(() => {
-    log.error('WebTransport failed to cleanup closed stream')
-  })
+class WebTransportStream extends AbstractStream {
+  private readonly writer: WritableStreamDefaultWriter<Uint8Array>
+  private readonly reader: ReadableStreamDefaultReader<Uint8Array>
 
-  let sinkSunk = false
-  const stream: Stream = {
-    id: streamId,
-    status: 'open',
-    writeStatus: 'ready',
-    readStatus: 'ready',
-    abort (err: Error) {
-      if (!writerClosed) {
-        writer.abort(err)
-          .catch(err => {
-            log.error('could not abort stream', err)
-          })
-        writerClosed = true
-      }
-      readerClosed = true
-
-      this.status = 'aborted'
-      this.writeStatus = 'closed'
-      this.readStatus = 'closed'
-
-      this.timeline.reset =
-        this.timeline.close =
-        this.timeline.closeRead =
-        this.timeline.closeWrite = Date.now()
-
-      cleanupStreamFromActiveStreams()
-    },
-    async close (options?: AbortOptions) {
-      this.status = 'closing'
-
-      await Promise.all([
-        stream.closeRead(options),
-        stream.closeWrite(options)
-      ])
-
-      cleanupStreamFromActiveStreams()
-
-      this.status = 'closed'
-      this.timeline.close = Date.now()
-    },
-
-    async closeRead (options?: AbortOptions) {
-      if (!readerClosed) {
-        this.readStatus = 'closing'
-
-        try {
-          await reader.cancel()
-        } catch (err: any) {
-          if (err.toString().includes('RESET_STREAM') === true) {
-            writerClosed = true
-          }
-        }
+  constructor (init: WebTransportStreamInit) {
+    super(init)
 
-        this.timeline.closeRead = Date.now()
-        this.readStatus = 'closed'
+    this.writer = init.bidiStream.writable.getWriter()
+    this.reader = init.bidiStream.readable.getReader()
 
-        readerClosed = true
-      }
+    Promise.resolve().then(async () => {
+      while (true) {
+        const result = await this.reader.read()
 
-      if (writerClosed) {
-        cleanupStreamFromActiveStreams()
-      }
-    },
+        if (result.done) {
+          init.log('remote closed write')
+          return
+        }
 
-    async closeWrite (options?: AbortOptions) {
-      if (!writerClosed) {
-        writerClosed = true
+        if (result.value != null) {
+          this.sourcePush(new Uint8ArrayList(result.value))
+        }
```

Comment:
Can value exist and the read be done? we should check for value first.

Reply:
Not exactly - if `value` exists and `done` is false, then `value` is the return value of the iterator.

See the iterator protocol docs: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols

Reply:
ah. then i've incorrectly implemented iterators previously, or consumed some incorrectly implemented ones, because i've seen a sequence value and a done=true... good to know that's not a correct implementation.

Reply:
😬